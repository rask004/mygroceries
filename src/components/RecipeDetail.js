import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {OrderedList, OrderedListAddButton} from '../ui/lists';
import {AddForm, EditableTextItem} from '../ui/forms';
import {FaPencilAlt, FaTrash, FaSave, FaRegTimesCircle} from 'react-icons/fa';


class InstructionsSection extends Component {
    constructor(props) {
        super(props);

        this.state = {
            newInstruction: "",
            editingInstruction: -1,
            editedInstructionText: "",
        }

        this.onChange = this.onChange.bind(this);
        this.newInstruction = this.newInstruction.bind(this);
        this.handleSaveEditedInstruction = this.handleSaveEditedInstruction.bind(this);
        this.handleCancelEditingInstruction = this.handleCancelEditingInstruction.bind(this);
        this.handleToggleEditingInstruction = this.handleToggleEditingInstruction.bind(this);
    }

    newInstruction(event) {
        event.preventDefault();
        const instruction = this.state.newInstruction;
        this.props.addInstruction(instruction);
        this.setState({newInstruction: ""});
    }

    onChange(event) {
        const target = event.target;
        const name = target.name;
        const value = target.value;
        this.setState({[name] : value});
    }

    handleSaveEditedInstruction (event, index) {
        event.preventDefault();
        const text = this.state.editedInstructionText;
        this.props.saveInstruction(index, text);
        this.setState({
            editingInstruction: -1,
            editedInstructionText: "",
        });
    }

    handleCancelEditingInstruction() {
        this.setState({editingInstruction: -1});
    }

    handleToggleEditingInstruction(index) {
        this.setState({
            editingInstruction: index,
            editedInstructionText: this.props.recipe.instructions[index],
        });
    }

    render() {
        const instructions = this.props.instructions.map((item, step) => { 
            return (
                <li key={step}>
                    {
                        this.state.editingInstruction === step ?
                        <form className="detail-instruction" onSubmit={(e) => this.handleSaveEditedInstruction(e, step)}>
                            <input value={this.state.editedInstructionText} onChange={this.onChange}
                                name="editedInstructionText" required placeholder="There must be an instruction here." />
                            <button type="submit"
                                ><FaSave/></button>
                            <button type="button" 
                                onClick={this.handleCancelEditingInstruction}><FaRegTimesCircle/></button>
                        </form> :
                        <div className="detail-instruction">
                            <div className="content">{item}</div>
                            <button type="button" onClick={() => this.handleToggleEditingInstruction(step)}><FaPencilAlt/></button>
                            <button type="button" onClick={() => this.props.removeInstruction(step)}><FaTrash/></button>
                        </div>
                    }
                </li>
            )
        });

        return (
            <div className="detail-instructions">
                <OrderedList 
                    title="Instructions"
                >
                    {instructions}
                </OrderedList>
                <AddForm addItem={this.newInstruction}
                    className={this.props.classNameAddform}>
                    <input name="newInstruction" required
                        value={this.state.newInstruction} 
                        onChange={this.onChange}
                        placeholder="new instruction..." />
                </AddForm>
            </div>
        );
    }
}

InstructionsSection.propTypes = {
    addInstruction: PropTypes.func.isRequired,
    saveInstruction: PropTypes.func.isRequired,
    removeInstruction: PropTypes.func.isRequired,
    instructions: PropTypes.array.isRequired,
}


class IngredientsSection extends Component {
    render() {
        const ingredients = this.props.ingredients.map((item, step) => { 
            return (
                <li key={item.id}>
                    <div className="content">{item.name}</div>
                    <div className="amount">{item.quantity}</div>
                    <div className="unit">{item.unit}</div>
                    <button type="button" 
                        onClick={() => this.props.removeIngredient(item.id)}>
                      <FaTrash/>
                    </button>
                </li>
            )
        });

        return (
            <OrderedListAddButton
                className="detail-ingredients"
                title="Ingredients"
                onClick={this.props.addIngredient}
                buttonTitle="Add Ingredient..."
            >
                {ingredients}
            </OrderedListAddButton>
        );
    }
}

IngredientsSection.propTypes = {
    addIngredient: PropTypes.func.isRequired,
    removeIngredient: PropTypes.func.isRequired,
    ingredients: PropTypes.array.isRequired,
}


class TitleSection extends Component {
    render() {
        return (
            <EditableTextItem 
                className="detail-title"
                updateText={this.props.updateTitle}
                text={this.props.title}
                placeholder="Recipe Title"
            />
        );
    }
}


export {InstructionsSection, IngredientsSection, TitleSection};


