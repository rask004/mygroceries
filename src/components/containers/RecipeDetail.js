import React, { Component } from 'react';
import {connect} from 'react-redux';
import {withRouter} from 'react-router-dom';
import PropTypes from 'prop-types';
import ConnectedNewIngredientModal from './NewIngredientModal';
import InstructionsList from './RecipeInstructions';
import IngredientsList from './RecipeIngredients';
import RecipeTitle from './RecipeTitle';
import {updateRecipe, addProduct} from '../../store/actions';
import Card from '@material-ui/core/Card';
import '../../css/recipedetail.css';


const NewIngredientModal = withRouter(ConnectedNewIngredientModal);


class RecipeDetail extends Component {

    getRecipe() {
        const id = Number(this.props.match.params.id);
        const recipe = this.props.recipes.find(item => item.id === id);
        return recipe;
    }

    constructor(props) {
        super(props);
        // FIXME id should be prop, with default value of undefined.
        // if undefined, use props.match.params.id if available.
        // if no props.match, use -1 (no recipe selected);
        const id = Number(props.match.params.id);
        const recipe = props.recipes.find(item => item.id === id);
        let value;
        value = recipe ?
            recipe.title : "";
        
        this.state = {
            newTitle: value,
            showNewIngredientModal: false,
        }

        this.handleSubmitNewInstruction = this.handleSubmitNewInstruction.bind(this);
        this.handleSubmitNewTitle = this.handleSubmitNewTitle.bind(this);
        this.handleRemoveInstruction = this.handleRemoveInstruction.bind(this);
        this.handleSaveInstruction = this.handleSaveInstruction.bind(this);
        this.toggleNewIngredientModal = this.toggleNewIngredientModal.bind(this);
        this.handleSubmitIngredient = this.handleSubmitIngredient.bind(this);
        this.handleRemoveIngredient = this.handleRemoveIngredient.bind(this);
        this.getRecipe = this.getRecipe.bind(this);
    }

    toggleNewIngredientModal() {
        this.setState({showNewIngredientModal: !this.state.showNewIngredientModal});
    }

    handleSaveInstruction(index, text) {
        const recipe = this.getRecipe();
        recipe.instructions[index] = text;
        this.props.onUpdateRecipe(recipe);
    }

    handleSubmitNewInstruction(text) {
        const recipe = this.getRecipe();
        recipe.instructions.push(text);
        this.props.onUpdateRecipe(recipe);
    }

    handleRemoveInstruction(index) {
        const recipe = this.getRecipe();
        recipe.instructions.splice(index, 1);
        this.props.onUpdateRecipe(recipe);
    }

    handleSubmitNewTitle(title) {
        const recipe = this.getRecipe();
        recipe.title = title;
        this.props.onUpdateRecipe(recipe);
    }

    handleRemoveIngredient(id) {
        const recipe = this.getRecipe();
        const ingredients = recipe.ingredients.slice().filter(item => item.id !== id);
        recipe.ingredients = ingredients;
        this.props.onUpdateRecipe(recipe);
    }

    handleSubmitIngredient(ingredient, quantity, unit) {
        let ingredientWasAdded = false;

        // first handle if new product, and get ingredient id
        const matchProduct = product => 
                product.name === ingredient.name &&
                product.coo === ingredient.coo &&
                product.brand === ingredient.brand &&
                product.category === ingredient.category &&
                product.subcategory === ingredient.subcategory;

        if(this.props.products.some(matchProduct)) {
            const products = this.props.products.slice().filter(matchProduct);
            ingredient.id = products[0].id;

        } else {
            const ids = this.props.products.map(item => item.id);
            ingredient.id = Math.max(...ids) + 1;
            this.props.onAddProduct(ingredient);
        }

        // now handle updating recipe, and informing user of any issues arising.

        const recipe = this.getRecipe();
        const existingIngredients = recipe.ingredients.slice();
        
        const existingIngredient = existingIngredients.find(x => x.id === ingredient.id);
        if (existingIngredient === undefined) {
            const recipeIngredient = {
                id: ingredient.id, 
                quantity: quantity,
                unit: unit,
            }
            existingIngredients.push(recipeIngredient);
            recipe.ingredients = existingIngredients;
            this.props.onUpdateRecipe(recipe);
            this.setState({showNewIngredientModal: !this.state.showNewIngredientModal});
            ingredientWasAdded = true;
        } else if ( existingIngredient.unit === unit) {
            const ingredientIndex = existingIngredients.findIndex(item => item.id === existingIngredient.id);
            const count = Number(existingIngredient.quantity) + Number(quantity);
            existingIngredients[ingredientIndex].quantity = count.toString();
            recipe.ingredients = existingIngredients;
            this.props.onUpdateRecipe(recipe);
            ingredientWasAdded = true;
            alert("Ingredient already exists. The quantity of the existing ingredient will be updated.");
            this.setState({showNewIngredientModal: !this.state.showNewIngredientModal});
        } else {
            alert("Error: You tried adding an existing ingredient with a different unit, which is not allowed. The existing ingredient has units of '" + existingIngredient.unit + "'");
        }

        return ingredientWasAdded;
    }

    render() {
        const recipe = this.getRecipe();

        let ingredients;
        let instructions;

        if (recipe) {
            instructions = recipe.instructions.slice();
            ingredients = recipe.ingredients.map((item) => {
                const product = this.props.products.find(x => x.id === item.id);
                if(product !== undefined) {
                    item.name = product.name;
                }
                return (
                    item
                );
            });
        }

        return (
            <React.Fragment>
                <Card className="recipe-detail">
                    {
                        recipe ?
                        <React.Fragment>
                            <RecipeTitle title={recipe.title} updateTitle={this.handleSubmitNewTitle}  />

                            <IngredientsList 
                                ingredients={ingredients}
                                addIngredient={this.toggleNewIngredientModal}
                                removeIngredient={this.handleRemoveIngredient}
                            />

                            <InstructionsList 
                                instructions={instructions}
                                saveInstruction={this.handleSaveInstruction}
                                removeInstruction={this.handleRemoveInstruction}
                                addInstruction={this.handleSubmitNewInstruction}
                            />

                        </React.Fragment> :
                        <div>
                            <p>No recipe is selected.</p>
                        </div>
                    }
                </Card>

                <NewIngredientModal show={this.state.showNewIngredientModal} 
                onClose={this.toggleNewIngredientModal}
                onAddIngredient={this.handleSubmitIngredient}
                />

            </React.Fragment>
        );
    }
}

RecipeDetail.propTypes = {
    onUpdateRecipe: PropTypes.func.isRequired,
    onAddProduct: PropTypes.func.isRequired,
    recipes: PropTypes.array.isRequired,
    products: PropTypes.array.isRequired 
}


const mapStateToProps = state => {
	return {
        recipes: state.recipes,
        products: state.products
    }
}

const mapDispatchToProps = dispatch => {
	return {
		onUpdateRecipe(recipe) {
			dispatch(
                updateRecipe(recipe)
            );
        },
        onAddProduct(product) {
            dispatch(
                addProduct(product)
            );
        }
	}
}

export {RecipeDetail};

export default connect(mapStateToProps, mapDispatchToProps)(RecipeDetail);