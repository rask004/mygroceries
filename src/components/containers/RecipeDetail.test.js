import React from 'react';
import {create} from 'react-test-renderer';

import {RecipeDetail} from './RecipeDetail';

import sampleData from '../../data/initialState.json';


const products = sampleData.products;
const recipes = sampleData.recipes;

const detail = <RecipeDetail 
        products={products}
        recipes={recipes}
        onUpdateRecipe={jest.fn()}
        onAddProduct={jest.fn()}
    />


describe('RecipeDetail Component render tests', () => {

    test('renders correctly', () => {
        const component = create(detail);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
