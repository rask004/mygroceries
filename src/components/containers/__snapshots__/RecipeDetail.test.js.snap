// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeDetail Component render tests renders correctly 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1 recipe-detail"
>
  <h1
    className="MuiTypography-root-29 MuiTypography-display1-33 MuiTypography-gutterBottom-56"
  >
    <div
      className="MuiGrid-container-65 detail-title"
    >
      <span
        className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-md-3-124 detail-content-title"
      >
        test recipe
      </span>
      <span
        className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-md-1-122"
      >
        <button
          className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-toggle"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-163"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-241"
          />
        </button>
      </span>
    </div>
  </h1>
  <div
    className="detail-ingredients"
  >
    <h2
      className="MuiTypography-root-29 MuiTypography-title-35 MuiTypography-gutterBottom-56"
    >
      Ingredients
    </h2>
    <ul
      className="MuiList-root-191 MuiList-padding-192"
    >
      <li>
        <span
          className="MuiTypography-root-29 MuiTypography-body1-38"
        >
          <div
            className="MuiGrid-container-65"
          >
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-4-125 MuiGrid-grid-lg-5-140 detail-content"
            >
              apple
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-3-124 MuiGrid-grid-lg-4-139 detail-content"
            />
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 MuiGrid-grid-lg-1-136 detail-content"
            >
              2
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 MuiGrid-grid-lg-1-136 detail-content"
            >
              each
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-1-122 MuiGrid-grid-lg-1-136"
            >
              <button
                className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-remove"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-163"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root-241"
                />
              </button>
            </span>
          </div>
        </span>
      </li>
      <li>
        <span
          className="MuiTypography-root-29 MuiTypography-body1-38"
        >
          <div
            className="MuiGrid-container-65"
          >
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-4-125 MuiGrid-grid-lg-5-140 detail-content"
            >
              orange
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-3-124 MuiGrid-grid-lg-4-139 detail-content"
            />
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 MuiGrid-grid-lg-1-136 detail-content"
            >
              3
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 MuiGrid-grid-lg-1-136 detail-content"
            >
              each
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-1-122 MuiGrid-grid-lg-1-136"
            >
              <button
                className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-remove"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-163"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root-241"
                />
              </button>
            </span>
          </div>
        </span>
      </li>
    </ul>
    <button
      className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-add"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-163"
      >
        Add Ingredient...
      </span>
      <span
        className="MuiTouchRipple-root-241"
      />
    </button>
  </div>
  <div
    className="detail-instructions"
  >
    <h2
      className="MuiTypography-root-29 MuiTypography-title-35 MuiTypography-gutterBottom-56 detail-subtitle"
    >
      Instructions
    </h2>
    <ol
      className="MuiList-root-191 MuiList-padding-192"
    >
      <li>
        <span
          className="MuiTypography-root-29 MuiTypography-body1-38"
        >
          <div
            className="MuiGrid-container-65"
          >
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-12-119 MuiGrid-grid-md-9-130 MuiGrid-grid-lg-10-145 detail-content"
            >
              eat the apples.
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 MuiGrid-grid-lg-1-136"
            >
              <button
                className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-toggle"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-163"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root-241"
                />
              </button>
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-1-122 MuiGrid-grid-lg-1-136"
            >
              <button
                className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-remove"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-163"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root-241"
                />
              </button>
            </span>
          </div>
        </span>
      </li>
      <li>
        <span
          className="MuiTypography-root-29 MuiTypography-body1-38"
        >
          <div
            className="MuiGrid-container-65"
          >
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-12-119 MuiGrid-grid-md-9-130 MuiGrid-grid-lg-10-145 detail-content"
            >
              eat the oranges.
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 MuiGrid-grid-lg-1-136"
            >
              <button
                className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-toggle"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-163"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root-241"
                />
              </button>
            </span>
            <span
              className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-1-122 MuiGrid-grid-lg-1-136"
            >
              <button
                className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-remove"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-163"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root-241"
                />
              </button>
            </span>
          </div>
        </span>
      </li>
    </ol>
    <form
      className="instruction-add"
      onSubmit={[Function]}
    >
      <button
        className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-add"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="submit"
      >
        <span
          className="MuiButton-label-163"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-241"
        />
      </button>
      <div
        className="MuiInputBase-root-208 MuiInput-root-195 MuiInput-underline-199 input-new-instruction"
        onClick={[Function]}
        required={true}
      >
        <input
          className="MuiInputBase-input-218 MuiInput-input-203"
          name="newItemName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="New instruction..."
          required={true}
          type="text"
          value=""
        />
      </div>
    </form>
  </div>
</div>
`;
